clingox/backends.py

```diff
@@ -0,0 +1,361 @@
+        '''
+        Add a disjuntive or choice rule to the underline backend.
+
+        Parameters
+        ----------
+        head : Iterable[Symbol]
+            The program atoms forming the rule head.
+        pos_body : Iterable[Symbol]=()
+            The program symbols forming the positive body of the rule
+        neg_body : Iterable[Symbol]=()
+            The program symbols forming the negated body of the rule
+        choice : bool=False
+            Whether to add a disjunctive or choice rule.
+
+        Returns
+        -------
+        None
+
+        Notes
+        -----
+        Integrity constraints and normal rules can be added by using an empty or
+        singleton head list, respectively.
+        '''
+        head_ = (self._add_symbols(head))
+        body = chain(self._add_symbols(pos_body), self._add_negated_symbols(neg_body))
+        return self.backend.add_rule(head_, body, choice)
+
+    def add_weight_rule(self, head: Iterable[Symbol], lower: int, pos_body: Iterable[Tuple[Symbol, int]], neg_body: Iterable[Tuple[Symbol, int]], *, choice: bool = False) -> None:
+        '''
+        Add a disjuntive or choice rule with one weight constraint with a lower bound
+        in the body to the underline backend.
+
+        Parameters
+        ----------
+        head : Iterable[int]
+            The program atoms forming the rule head.
+        lower : int
+            The lower bound.
+        pos_body : Iterable[Tuple[Symbol,int]]
+            The pairs of program symbols and weights forming the elements of the
+            positive body of the weight constraint.
+        neg_body : Iterable[Tuple[Symbol,int]]
+            The pairs of program symbols and weights forming the elements of the
+            negative body of the weight constraint.
+        choice : bool=False
+            Whether to add a disjunctive or choice rule.
+
+        Returns
+        -------
+        None
+        '''
+        head_ = (self._add_symbols(head))
+        body = chain(self._add_symbols_weights(pos_body), self._add_negated_symbols_weights(neg_body))
+        return self.backend.add_weight_rule(head_, lower, body, choice)
+
+    def _add_symbols(self, symbols: Iterable[Symbol]) -> Iterable[int]:
+        '''
+        Return a fresh program atom or the atom associated with each symbol in `symbols`.
+
+        If the given symbol does not exist in the atom base, it is added first. Such
+        atoms will be used in subequents calls to ground for instantiation.
+
+        Parameters
+        ----------
+        symbols : Iterable[Symbol]
+            The symbols associated with the atoms.
+
+        Returns
+        -------
+        Iterable[int]
+            The program atoms representing the atoms.
+        '''
+        return (self.backend.add_atom(symbol) for symbol in symbols)
+
+    def _add_negated_symbols(self, symbols: Iterable[Symbol]):
```

Comment:
I would combine with the previous function to keep the code compact. `_add_symbols(self, sign: bool, symbols: Iterable[Symbol])` should do.